README.md
# CookbookV3
# GUI Cookbook program written in Python and Tk

This project is to act as a repository for my cookbook project.  It is written using Python 3.7.4 and tkinter.  The GUI designer is Page written by Don Rozenberg.  Page is Free Open Source software.  The website for Page is [https://sourceforge.net/projects/page/](https://sourceforge.net/projects/page/)

My cookbook program started out in the early 2000's as a recipe database for Android tablets written in B4A.  Version 2 was simply a re-write under Python using Page generated GUIs.  This is version 3 and is still in early development.

Python version 3.7x must be used for this project.

### NOTICE: I have made major changes to the code on 31 January, 2020.  Please make a copy of your recipe database and recipeimages folder.  I am currently working on a migration tool that will allow you to import any recipes that you might have added.  You will likely run into a bug or two hundred.  Please let me know if you find any issues.

## Screenshots (new ones will be uploaded soon)...

The main screen
![The main cookbook screen](./assets/main.png)

The scraper screen
![The scraper screen](./assets/scraper.png)

The editor/entry screen

**coming soon**

More images will become available as development continues.

## Required libraries
At the moment, there are three libraries that will need to be installed before you attempt to run the program(s).  They are:

* Pillow
* requests
* recipe_scrapers
* tkcolorpicker
* _*ttkthemes >= 3.0.0 (new requirement as of 11 February 2020)*_


All of these libraries can be installed using the Python pip installer.  As an example, to install the requests library, for Python 2.x you would use:
```bash
$ pip install requests
```
or
```bash
pip3 install requests
```
if you are using Python 3.x (which at this time you _*SHOULD*_ be using anyway).

## Sourcecode considerations

I can, without fear of contradiction, state that my coding style is very old school and more verbose that it "needs" to be.  Basically, I feel that I am a teacher, since I share my source code freely with anyone.  Therefore, I write as if anyone who reads my code is using it for learning purposes.  So I tend to keep the cryptic programming at a level of almost non-existance.  If you feel you don't like my coding style, that's alright.  Others seem to, so you must be in the minority. ;o}

I should put a quick note here about the number of files that make up my program.  When Page generates a GUI, there are three files that are created.  Assuming that your project is named "Project", it would create the following three files:

* Project.tcl (The Page .tcl file)
* Project.py  (What I call the GUI file)
* Project_support.py (where all the modifiable code is placed.)

It is _*NEVER*_ a good idea to modify either of the first two files by hand.  If you do so, there is a very good chance that your program will crash and you will be left with an unusable set of source code.  This is how Page works, and I like it.  When you want to see actual Python source code in any of my projects that use Page, the first place to look at is the *_support.py file.  Earlier, I alluded to the fact that my code is verbose and it's true.  Another habit of mine is that I comment the heck out of my code.  Again, it's because other people learn from my code, I write articles where I teach people about Python programming, and I'm old and have a hard time remembering what I was thinking when I wrote that code.  If you fork or clone or otherwise copy my code, that's fine.  Once it's on your machine, you can remove the comments if you feel moved to do so.

_*One final thing.  The .tcl files that Page creates are not compatible with any current versions of Page.  The project was designed with Page 5b, which as of 31 January 2020, is still in development.  Any previous version of Page will refuse to open these files.*_

## Acknowledgments

This project would not be possible if it weren't for the efforts of many other people, who have kindly and freely provided their source code.

*  Don Rozenberg - Author of __Page__, the GUI designer for Python - https://sourceforge.net/projects/page/
*  The authors and contributers of __recipe_scrapers__ - https://github.com/hhursev/recipe-scrapers
*  All maintainers of __Python__ - https://python.org

### About the author

My name is _*Greg Walters*_.  I have been involved with computers and computer programming since 1972.
I am well versed in Visual Basic, Visual Basic .NET, Python, and SQL, and is an
accomplished user of MySQL, SQLite, Microsoft SQL Server, Oracle, C++, Delphi,
Modula-2, Pascal, C, 80x86 Assembler, COBOL, and Fortran. I have been a programming trainer and has trained numerous people on many pieces of computer software, including MySQL, Open Database Connectivity, Quattro Pro, Corel Draw!, Paradox, Microsoft Word, Excel, DOS, Windows 3.11, Windows for Workgroups, Windows 95, Windows NT, Windows
2000, Windows XP, and Linux. I am semi-retired and has written over 100 articles for the Full Circle magazine. I am open to working as a freelancer on various projects.  I also have authored one book solo, co-authored another and am currently acting as a freelance Technical Reviewer for *Packt Publishing* having done 7 books to date.

As you can tell, I have been around a long time.

Please note:  This readme file is a living document, as is the source code, and subject to being changed at any time.  Usually in the early mornings when I haven't had enought coffee.

Thank you for stopping by.  I hope you enjoy the code and the program.

*greg*

main/changelog.txt
Changelog.txt

02 January, 2021 -
Scraper - Added Duplicate Recipe Check in Scraper - Need to same for New recipe in Editor Form

    # Check for duplicate
    sql = f"SELECT RecipeText, RecipeSource FROM recipes WHERE RecipeText = {quote(sRecipeTitle.get())} and RecipeSource = {quote(EntryWebsite.get())}"
    resp = list(cursor.execute(sql))
    if len(resp) > 0:
        title='Save Data'
        msg = 'This recipe already exists in the database!'
        messagebox.showwarning(title,msg)
        busyEnd()
    else:
        ...
----------------------

21 December, 2020 -
Main - 3.4.9.0
Fixed support for Favorite recipes.
Database currently has over 400 recipes.
Changed print function to call Firefox directly due to changes in Google Chrome affecting printing.
In main window, if you click on the recipe source hyperlink (blue), link is sent to Firefox to show recipe source.
Cleaned up some of the code.
Still have lots to do.
---------------------

24 May, 2020 -
Main - Began Adding support for favorites
Database - Began Adding support for favorites
-------------------

1 April 2020 -
Main - 3.4.8.4
     - Started support for tips form.  Currently shows "under construction" messagebox.

-------------------
31 March -
Database - 0.2
         - Added support for tips datatable
         - Added support for system table (saves version number

12 March -
Print - 0.1.2 
      - Fixed bug where recipe title was not printing at top
Scraper - 0.4.0
        - Changed form title to "Greg's Recipe Website Scraper"

------------------

3 March -
Main - 3.4.8.3
     - 
Config - 0.1.3
       - Added ShowAllRecords default to config.
       - Added combo box for default theme.

------------------

2 March - 
Main - 3.4.8.2
     - Fixed bug I-8

-----------------

1 March - 

Main - 3.4.8.1
     - Fixed Treeview focus issue


-----------------

1 March - 

Main - 3.4.8
     - Added logic to return treeview and form to recipe selected when new recipe was entered or recipe was scraped.
     - Still ToDo - Treeview still not focusing on returned recipe name, but is being "selected"

Scraper - 0.3.9
        - Support for return to new recipe in Main
-----------------

28 February - 

Main - 3.4.7
     - Added logic to return treeview and form to recipe selected when edit was clicked.

TODO - Add logic to do same when returning from new recipe entry and scape

Scraper - 0.3.8
        - fixed issue that was keeping image from being retrieved
TODO    - Verify and/or fix when response is 400

Config - 0.0.1
       - Added first blush of config utility and support to Main

-----------------
27 February -
Editor - 0.5.6
       - rearranged some statements in write_to_db
Scraper - 0.3.7
        - minor fixes
Main    - 3.4.6
        - minor fixes
-----------------

26 February -

Main - Added support for config file
Scraper - Added support for scraper sites that require agent 
Editor - (in progress) Modified code to support a default path for image rather than /home
TODO - Modify configuration utility to support the config file.

-----------------

8 February 0600

Main - Version 3.4.4
       Modified form - replaced label widget with message widget for Source to better support longer URLs

----------------
3 February 1100

Main - 3.4.2
     - Fixed small problem with running main program from a folder not within main folder.

---------------
31 January 1830
DB Utility - Version 0.1.4
           - Minor fix to allow standalone execution
---------------
31 January 1815

DB Utility - Version 0.1.3
           - Added ability to add new categories

---------------

31 January 1730

Some minor changes to UI for Main and Scraper.

---------------
31 January 1215
Multiple fixes for images/database caused by new file structure

---------------
31 January 0600
Updated README.md

---------------
31 January 

Rewrote multiple parts of code in all modules to support new structure

---------------
29 January Evening

* Main - 3.4.0
       - Added config button and db maint button
* DBUtils - 0.1.1
          - Initial release - NO migrate function


---------------
29 January 1215

* Several small tweeks
* Many new recipes
* Beginning work on database utility form

---------------
29 January

* Scraper - Version 0.3.3
          - Added routine to check URL domain against supported scrapers before attempting to scrape.
          - Added beginnings of "darker" theme

* Main - Version 3.2.2
       - Removed code that made form on top of all others upon deiconify.
       - Fixed CurrentID not set on startup
       - Added beginnings of "darker" theme
       - Added support for viewing active records only or all records

* Editor - Version 0.5.4
         - Added beginnings of "darker" theme 
* Database - categories - Added chicken, turkey, duck, Latin American, Side Dish, Marinade

---------------
28 January 1130

All Modules - minor code cleanup
Main - Version 3.2.1
     - Removed button image assignment in Page and make assignments in code

Entered bug I-6 (images not scraping properly on SOME supported web site recipe pages)
Fixed bug I-5 
Created workaround for bug I-6

---------------
28 January 0400

* Main - Version 3.2.0
       - Added support for printing recipe to HTML.
       - Minor cleanup
* Scraper - Version 0.3.0
          - Minor changes to try to minimize web pages that don't properly return images.

---------------
27 January 0745

* Added text file showing currently supported recipe websites for scraping (CurrentlySupportedRecipeSitesForScraping.txt)
* Editor - Version 0.5.3
         - Added ability to edit and/or delete single item in ingredients listbox
         - Might change '<<ListboxSelect>>' to a double click to keep from accidental triggering.
         - Disabled active checkbox and recordnumber labels if in new entry mode

---------------
27 January 0630
* Editor - Version 0.4.1
* Cleared bug I-4 Editor
  Also fixed same issue in Scraper
* Cleaned up init function

main/formConfig.py
#! /usr/bin/env python
#  -*- coding: utf-8 -*-
#
# GUI module generated by PAGE version 5.0.2
#  in conjunction with Tcl version 8.6
#    Mar 03, 2020 05:52:07 AM CST  platform: Linux

import sys

try:
    import Tkinter as tk
except ImportError:
    import tkinter as tk

try:
    import ttk
    py3 = False
except ImportError:
    import tkinter.ttk as ttk
    py3 = True

import formConfig_support

def vp_start_gui():
    '''Starting point when module is the main routine.'''
    global val, w, root
    root = tk.Tk()
    formConfig_support.set_Tk_var()
    top = formConfig (root)
    formConfig_support.init(root, top)
    root.mainloop()

w = None
def create_formConfig(rt, *args, **kwargs):
    '''Starting point when module is imported by another module.
       Correct form of call: 'create_formConfig(root, *args, **kwargs)' .'''
    global w, w_win, root
    #rt = root
    root = rt
    w = tk.Toplevel (root)
    formConfig_support.set_Tk_var()
    top = formConfig (w)
    formConfig_support.init(w, top, *args, **kwargs)
    return (w, top)

def destroy_formConfig():
    global w
    w.destroy()
    w = None

class formConfig:
    def __init__(self, top=None):
        '''This class configures and populates the toplevel window.
           top is the toplevel containing window.'''
        _bgcolor = '#d9d9d9'  # X11 color: 'gray85'
        _fgcolor = '#000000'  # X11 color: 'black'
        _compcolor = '#d9d9d9' # X11 color: 'gray85'
        _ana1color = '#d9d9d9' # X11 color: 'gray85'
        _ana2color = '#ececec' # Closest X11 color: 'gray92'
        font17 = "-family {Ubuntu} -size 10 -weight bold"
        font18 = "-family {Ubuntu} -size 10 -weight bold"
        self.style = ttk.Style()
        if sys.platform == "win32":
            self.style.theme_use('winnative')
        self.style.configure('.',background=_bgcolor)
        self.style.configure('.',foreground=_fgcolor)
        self.style.configure('.',font="TkDefaultFont")
        self.style.map('.',background=
            [('selected', _compcolor), ('active',_ana2color)])

        top.geometry("758x544+486+266")
        top.minsize(1, 1)
        top.maxsize(1905, 1050)
        top.resizable(1, 1)
        top.title("Configuration Utility")
        top.configure(highlightcolor="black")

        self.Frame1 = tk.Frame(top)
        self.Frame1.place(x=2, y=46, height=495, width=755)
        self.Frame1.configure(relief='groove')
        self.Frame1.configure(borderwidth="2")
        self.Frame1.configure(relief="groove")

        self.style.configure('TNotebook.Tab', background=_bgcolor)
        self.style.configure('TNotebook.Tab', foreground=_fgcolor)
        self.style.map('TNotebook.Tab', background=
            [('selected', _compcolor), ('active',_ana2color)])
        self.TNotebook1 = ttk.Notebook(self.Frame1)
        self.TNotebook1.place(x=5, y=5, height=485, width=745)
        self.TNotebook1.configure(takefocus="")
        self.TNotebook1_t1 = tk.Frame(self.TNotebook1)
        self.TNotebook1.add(self.TNotebook1_t1, padding=3)
        self.TNotebook1.tab(0, text="General Configuration", compound="left"
                ,underline="-1", )
        self.TNotebook1_t2 = tk.Frame(self.TNotebook1)
        self.TNotebook1.add(self.TNotebook1_t2, padding=3)
        self.TNotebook1.tab(1, text="Themes",compound="left",underline="-1",)

        self.Label1 = tk.Label(self.TNotebook1_t1)
        self.Label1.place(x=20, y=100,  height=19, width=186)
        self.Label1.configure(activebackground="#f9f9f9")
        self.Label1.configure(anchor='e')
        self.Label1.configure(font=font18)
        self.Label1.configure(text='''Future:''')

        self.Message1 = tk.Message(self.TNotebook1_t1)
        self.Message1.place(x=10, y=44, height=43, width=235)
        self.Message1.configure(font=font17)
        self.Message1.configure(padx="3")
        self.Message1.configure(pady="3")
        self.Message1.configure(text='''Default source path for images for manual addition of images''')
        self.Message1.configure(width=235)

        self.Entry1 = tk.Entry(self.TNotebook1_t1)
        self.Entry1.place(x=250, y=50, height=31, width=246)
        self.Entry1.configure(background="white")
        self.Entry1.configure(font="-family {Ubuntu} -size 10 -weight bold")
        self.Entry1.configure(selectbackground="#c4c4c4")
        self.Entry1.configure(textvariable=formConfig_support.defImagePath)

        self.btnFilePath = tk.Button(self.TNotebook1_t1)
        self.btnFilePath.place(x=500, y=50,  height=29, width=40)
        self.btnFilePath.configure(activebackground="#f9f9f9")
        self.btnFilePath.configure(command=formConfig_support.on_btnFilePath)
        self.btnFilePath.configure(text='''...''')

        self.chkAllRecs = tk.Checkbutton(self.TNotebook1_t1)
        self.chkAllRecs.place(x=245, y=139, height=21, width=63)
        self.chkAllRecs.configure(anchor='w')
        self.chkAllRecs.configure(justify='left')
        self.chkAllRecs.configure(variable=formConfig_support.chkShowAllRecs)

        self.Label2 = tk.Label(self.TNotebook1_t1)
        self.Label2.place(x=20, y=140,  height=19, width=216)
        self.Label2.configure(anchor='e')
        self.Label2.configure(font=font18)
        self.Label2.configure(text='''Show all records by default:''')

        self.TCombobox1 = ttk.Combobox(self.TNotebook1_t2)
        self.TCombobox1.place(x=250, y=50, height=19, width=177)
        self.TCombobox1.configure(textvariable=formConfig_support.combobox)
        self.TCombobox1.configure(takefocus="")

        self.Label3 = tk.Label(self.TNotebook1_t2)
        self.Label3.place(x=20, y=50,  height=19, width=206)
        self.Label3.configure(anchor='e')
        self.Label3.configure(font=font18)
        self.Label3.configure(text='''Default Theme:''')

        self.Frame2 = tk.Frame(top)
        self.Frame2.place(x=2, y=2, height=45, width=755)
        self.Frame2.configure(relief='groove')
        self.Frame2.configure(borderwidth="2")
        self.Frame2.configure(relief="groove")

        self.btnExit = tk.Button(self.Frame2)
        self.btnExit.place(x=714, y=3,  height=35, width=35)
        self.btnExit.configure(activebackground="#f9f9f9")
        self.btnExit.configure(command=formConfig_support.on_btnExit)
        self.btnExit.configure(text='''Exit''')

if __name__ == '__main__':
    vp_start_gui()





